devolutions-session/src/dvc/task.rs

```diff
@@ -830,6 +841,49 @@ async fn handle_exec_error(dvc_tx: &WinapiSignaledSender<NowMessage<'static>>, s
     }
 }
 
+fn set_kbd_layout(layout: SetKbdLayoutOption) -> anyhow::Result<()> {
+    // SAFETY: No outstanding preconditions.
+    let foreground_window = unsafe { GetForegroundWindow() };
+
+    if foreground_window.is_invalid() {
+        bail!("Failed to get foreground window handle");
+    }
+
+    let locale = match layout {
+        SetKbdLayoutOption::Next => isize::try_from(HKL_NEXT).expect("HKL_NEXT fits into isize"),
+        SetKbdLayoutOption::Prev => isize::try_from(HKL_PREV).expect("HKL_PREV fits into isize"),
+        SetKbdLayoutOption::Specific(layout) => {
+            // IMPORTANT: WM_INPUTLANGCHANGEREQUEST message only respects low word of the layout
+            // (locale) and high word (device) should be zero. Non-zero high word could produce
+            // unexpected results (e.g. switching to next layout instead of specific one).
+            //
+            // Loading locale via LoadKeyboardLayoutW and passing it to WM_INPUTLANGCHANGEREQUEST
+            // will not work as expected (even if dozen of sources suggest this approach) and
+            // easily breaks with some layouts.
+
+            let hex = u32::from_str_radix(layout, 16)
+                .context("Invalid keyboard layout value")?;
+            // device == (hex >> 16) && 0xFFFF
+            let locale = u16::try_from(hex & 0xFFFF).expect("locale <= 0xFFFF");
+
+            isize::try_from(locale).expect("locale fits into isize")
+        }
+    };
+
+
+    /// SAFETY: hwnd is valid window handle, therefore `PostMessageW` is safe to call.
```

Comment:
nitpick: It’s always redundant to conclude with things like "therefore calling this unsafe function is safe" in SAFETY comments.

```suggestion
    /// SAFETY: hwnd is valid window handle.
```